### DISCUSSION TITLE
restic and backrest path change with every upgrade and need for permissions change

### DISCUSSION THREAD

**Discussion by amarendra** [garethgeorge/backrest#495] [Ideas]
I just started it using 2-3 days ago so still getting the hang of it but so far liking it. It's simple and it just works without getting in the way or how `restic` is configured.

However I noticed this when I upgraded from `1.5.0` to `1.5.1` that I had to give fresh permission to Photos/etc access, Full Disk Access, and also approval for `objective-see/LuLu` as well. Because apparently every version is referenced from its path which contains the  version number in it. 

Can it be avoided? Or is there already a way that can avoid it (maybe I am not aware of that)?

**Comment by garethgeorge** [comment]
Hey— thanks for filing this, hadn’t been aware macOS users were running into this. 

I think it could make sense for backrest to try to keep the path static and instead use the restic executable itself to check the version (via version command). 

Will macOS retain the full disk access permissions even if the binary checksum changes? Is the path the only thing that matters?

**Reply by amarendra** [reply]
So basically every single app across upgrades maintain those permissions so I believe that is possible. Just don't know how it is done.

**Reply by garethgeorge** [reply]
When you grant full disk access, are you granting it to the backrest binary or to restic? Unfortunately apple has no official docs on how the feature behaves, but iiuc it's supposed to propagate to child processes.

**Reply by amarendra** [reply]
I am granting it to `backrest` which is what the prompt was for. (I get promot for `restic` only when LuLu intercepts n/w calls, that is because that's the process making those n/w calls I guess).

> file access

(As of now `backrest+restic` is able to access all the files in my `include list` except `~/Library/Messages/chat.db`. Don't know what special permission is needed for this. 


> Unfortunately apple has no official docs on how the feature behaves

Yeah, that's classic Apple. 

When I go and see in `Privacy & Security` there are two section related to this:

`Files and Folders` which has 2 instances of `backrest`. Looks like the one greyed out is the one which was `1.5.0` and is orphaned there for some reason. So yeah keeping the path of binary static might help. It used to work for `restic cli`. I never had to redo the permissions either for Mac's privacy restrictions or for `LuLu`.

Let me remove the both `backrest` and then give fresh full disk access permission to the `1.5.1` backrest binary I will share here what happens. → This worked. `backrest` was able to access even `~/Library/Messages`. I verified it with snapshot browser (which is a neat killer feature by the way; thank you!).

(I do see `backrest` in `/opt/homebrew` which points to the latest `backrest` version - can that not made to work? Not sure how other apps from `homebrew` work, or how `restic cli` works)

**Reply by garethgeorge** [reply]
Hmm, I wonder if this is fundamentally a homebrew issue with the paths changing? That's annoying if the case -- using the included install script will keep the path static -- but isn't great because you miss out on automatic updates.

**Reply by amarendra** [reply]
> `/opt/homebrew`

Can you not have the service point to this path's binary? It anyway always points to the latest version, right? (But then I don't know how Mac treats permissions and symlinks). 

I guess might need to see how `restic` handles it. Because I needed to provide permissions only once across upgrades (iirc).

But Mac GUI apps anyway just move the latest `.app`  file (actually folder) to `/Applications/` and get done with it.

**Reply by garethgeorge** [reply]
Iiuc homebrew is creating a symlink? I think what I'm wondering is if the problem is that macOS is checking the actual location that the symlink resolves to, and that location might be changing after each upgrade?

**Reply by amarendra** [reply]
> Iiuc homebrew is creating a symlink?

I would certainly think so. 

I do see a conversation about it in `restic` issues from 2018 here - https://github.com/restic/restic/issues/2051#issuecomment-430746784. There might be more.

**Reply by amarendra** [reply]
Hi @garethgeorge. Just wanted to check whether this is on any roadmap or planned by any chance? Just curious.

**Reply by garethgeorge** [reply]
this is largely fixed on macOS. I'd add on however that you should actually only need to grant permissions (e.g. full disk access, etc) to backrest. I believe restic will inherit them.

Backrest will now try to use the system install of restic if it is up to date (e.g. this will often be the version provided by homebrew if you're using that package manager)

If it isn't, backrest now intelligently installs it's version of backrest at `~/.local/share/backrest/restic` (with no version suffix).

You can check which binary backrest uses by checking the logs ~/.local/share/backrest/processlogs/ and looking at the logs from the last backrest startup.

**Reply by amarendra** [reply]
So in 1.6.2 the logs indeed mention which version of restic is being used and in my case it is `~/.local/share/backrest/restic` indeed. I had to give permission to it separately - both network and disk.

However I still see backrest being listed as `/opt/homebrew/Cellar/backrest/1.6.2/bin/backrest` so the next version of backrest will  need permissions again. I have 3-4 `backrest` entires in Photos, iCloud, and Full Desk Access settings on Mac. Some of the older/residual ones I can remove, some I can't.

Then I removed Full Disk Access for every `backrest` entry and the backup still passed.

**Reply by amarendra** [reply]
Update notes:

The path still has to be added from Cellar directory. Adding this `/usr/local/bin/backrest` doesn't work. In fact on my Mac (via Homebrew) file/binary `/usr/local/bin/backrest` isn't there at all. 

However there is this symlink `/opt/homebrew/bin/backrest` which points to `/opt/homebrew/Cellar/backrest/1.7.3/bin/backrest`

**Comment by amarendra** [comment]
Hey @garethgeorge I thought I will finally tag you since I suspect you might not be receiving any notifications - hope it's not a bother. The steps to give permissions to backrest is still the same. I just upgraded to `1.8.0` and I had to go through this again. Also, the Photos access is tricky - because I was not able to "give" this permission, I can approve the popup when it runs next time and I see the popup and I notice it.

Also, the stale/orphaned permissions to old instances/versions of backrest still show their and after every upgrade one more is added (scr attached).

<img width="482" alt="Screenshot 2025-04-22 at 12 55 24 PM" src="https://github.com/user-attachments/assets/e09dde39-33ab-4211-adf8-25589c886c45" />

**Reply by garethgeorge** [reply]
Appreciate the ping, I've been traveling so catching up on threads. Unfortunately I've more or less concluded that I'm at a loss w/macOS because of the platform's _very_ opinionated stance on binary signing (and I say this as someone using a an m-series mac to develop backrest these days).

I think unfortunately my best recommendation at the moment for macOS users is to adopt a manual upgrade cadence if the permissions dialogs are a problem, homebrew is perhaps best avoided.

In the future I'd like to aim to add a "new version available" and "click to upgrade" self update feature to the Backrest UI which I'm thinking can ease some of this while still making upgrades an explicit thing where you have the opportunity to check that permissions are still OK after triggering it.

If you, or anyone, does find improvements I can make to make the upgrade process play better with macOS permissions I'm very open to suggestions here.

**Reply by amarendra** [reply]
Yeah I think it's an unsolvable problem then. And I guess adding a "click to upgrade" option would definitely help if that skips the re-grant-all-permissions part. Will keep an eye out for that release. I will also try to find ways to clear those stale names from settings.

Honestly it's Apple that is the problem here (and they are very deliberate about it - part of the control and stifle initiatives which they are using to tighten control on even the macOS now; e.g. their shenanigans with File Provider API) and to some extent brew's fundamental design choices. 

Thanks a lot for taking the time to respond.

> opinionated

That's being very generous to Apple :)

**Comment by amarendra** [comment]
Hi Gareth, bumping on this old issue yet again. So finally I went for this:

> I think unfortunately my best recommendation at the moment for macOS users is to adopt a manual upgrade cadence if the permissions dialogs are a problem, homebrew is perhaps best avoided.

and it was pretty smooth (hail Go binaries).

~So basically by "manual update" you mean - all I have to do is run the "install.sh" (from the latest stable darwin tar.gz) every time I notice an update (I am guessing it will be better than just replacing the binary myself as install script will take care of restarting the service etc, right?). This won't be a problem, I will keep checking even if there's no notification.~

~I hope this is what it is.~

Edit:
I saw in some other discussions post that essentially says something like - _manual install again is the way to manual upgrade_. So that is settled. I will just keep checking for backrest updates.

**One more doubt**: does manual install brings its own copy of `restic` and keeps it at `$HOME/.local/share/backrest/restic` and uses only this?
